include/sys/spa.h

```diff
@@ -155,6 +155,7 @@ _NOTE(CONSTCOND) } while (0)
 
 #define	SPA_COMPRESSBITS	7
 #define	SPA_VDEVBITS		24
+#define	SPA_COMPRESSMASK	((1U<<SPA_COMPRESSBITS)-1)
```

Comment:
thats half true. at some points the compression type field is or masked with a special level type. there are about 40 compression levels. since we later store just the compression type as you said, but the level in a separate field, we need to mask out the levels again. thats not my idea. its the way allan did it. he also explained why. gzip for instance has a new compression type for each level. but thats just 9. here we have alot of more levels. but anyway. i will fix the style issue

Reply:
The main reason for the mask is that we don't expect userspace to deal with the compression levels. the user does: `zfs set compress=zstd-10 dataset`, and then in zfs_ioctl.c we split that into compress= and compress_level=

The main reason to do this was to avoid odd situations like: compress=zstd compress_level=10 and then you do: zfs set compress=gzip, and compress_level=10 is invalid.

So I wanted a way to set the compression type and level at once. This was the solution suggested by Saso Kiselkov and Robert Mustacchi at the OpenZFS developer summit after my presentation. I am open to other approaches.

Reply:
i'm happy with it. i can't decide which approach is better or not. both are identical for me. no difference. the example for gzip isnt that good if course since we simply could ignore compress_level if gzip is set. but it would still leave a unused field which is odd. so since there is no dynamic way for adding or removing available options. using a single option for all compression properties makes sense and will not confuse users comparing to using  gzip